### ISSUE TITLE
FlurlHttpException helper creator 

### ISSUE THREAD

**Issue by francischie** [tmenier/Flurl#579]
I found it hard to create an instance FlurHttpException for my unit tests.  Took me a while to create a helper function for it. It would be nice to have some built-in way in the library to create this.  Or maybe there is a better way of doing this that I missed. 

```
 public static class UnitTestHelper
    {
        public static FlurlHttpException CreateFlurlHttpException(string httpContent = null, HttpStatusCode statusCode = HttpStatusCode.BadRequest)
        {
            var response = new HttpResponseMessage
            {
                StatusCode = statusCode,
                Content = new StringContent(httpContent ?? string.Empty)
            };
            return new FlurlHttpException
            (new Fixture().Build<FlurlCall>()
                .OmitAutoProperties()
                .With(a => a.Request, new FlurlRequest {Url = "http://dummy"})
                .With(a => a.HttpRequestMessage, new HttpRequestMessage())
                .With(a => a.HttpResponseMessage, response)
                .With(a => a.Response, new FlurlResponse(response))
                .Create());
        }
    }

```

I created a demo project for sample usage. I'm using AutoFixture by the way for creating this helper. 

https://github.com/francischie/FlurlHttpClientExceptionDemo/blob/main/FlurlHttpClientExceptionDemo.Tests/Services/SampleServiceTests.cs

**Comment by tmenier** [comment]
I guess I'm not quite sure what you're looking for as an improvement. `FlurlHttpException` requires only a `FlurlCall`, which has an empty ctor and a handful of read/write properties, none of which are strictly required. So I think what you're going to need to populate will depend entirely on what you're testing, and using AutoFixture seems like a good way to do this. Unless I'm missing something?

**Comment by francischie** [comment]
@tmenier , Passing FlurCall is not enough on my testing. I need to set up a few more properties to make it work. 

As I showed on the sample code. I need to set up Request, HttpRequestMessage, HttpResponseMessage, Response properties to make it work, and took me some time fiddling around. 

```
 return new FlurlHttpException
            (new Fixture().Build<FlurlCall>()
                .OmitAutoProperties()
                .With(a => a.Request, new FlurlRequest {Url = "http://dummy"})
                .With(a => a.HttpRequestMessage, new HttpRequestMessage())
                .With(a => a.HttpResponseMessage, response)
                .With(a => a.Response, new FlurlResponse(response))
                .Create());
```

**Comment by tmenier** [comment]
Strictly speaking, the minimum setup is this:
```
new FlurlHttpException(new FlurlCall());
```
If your test requires more, then that's relative to your test, not an absolute requirement. You've figured out what's needed for your specific scenario, coded it with the help of AutoFixture, and encapsulated it into a method that you can reuse. I think that's perfect.

Maybe it would help me understand if you explained what you envision a helper looking like? Just a constructor or method signature you have in mind is fine, doesn't need to be an implementation.

**Comment by francischie** [comment]
> Strictly speaking, the minimum setup is this:
> 
> ```
> new FlurlHttpException(new FlurlCall());
> ```
> 
> If your test requires more, then that's relative to your test, not an absolute requirement. You've figured out what's needed for your specific scenario, coded it with the help of AutoFixture, and encapsulated it into a method that you can reuse. I think that's perfect.
> 
> Maybe it would help me understand if you explained what you envision a helper looking like? Just a constructor or method signature you have in mind is fine, doesn't need to be an implementation.

That is where I actually started to create a simple FlurHttpException without AutoFixture. And it gave me these errors. Then I use AutoFixture to help me generate some needed properties. Still give a bunch of errors till I set the minimum properties to be setup. 

![image](https://user-images.githubusercontent.com/5061442/101395700-2a808580-387f-11eb-82db-4ef9811ef4c9.png)

The error might be a bug and should just allow checking the NullReferenceException then I don't even need this helper method. 

FYI, a demo project was created for this: https://github.com/francischie/FlurlHttpClientExceptionDemo

**Comment by francischie** [comment]
An overload for the `FlurlHttpException` parameter constructor accepting StatusCode and Content would be pretty straightforward helpful. 

Like 
```
var exception = new FlurlHttpException(500, "something went wrong (o some json format)");
```

**Comment by tmenier** [comment]
> The error might be a bug and should just allow checking the NullReferenceException then I don't even need this helper method.

Ah, yes, that's definitely something I'm willing to fix.

As for your suggested constructor, I don't want to make any promises but I'll give it some thought. I do think providing response details and not request details would result in something you'd never see thrown in a real situation though. But it's a start.

**Comment by francischie** [comment]
> As for your suggested constructor, I don't want to make any promises but I'll give it some thought. I do think providing response details and not request details would result in something you'd never see thrown in a real situation though. But it's a start.

Fair enough. I think fixing or allowing the creation of FlurlHttpException without passing anything should be good enough.  If users need to mock a request or response then they can set up it with the current existing constructor/property as you said

I just noticed this behavior while trying to do code coverage on our try-catch block that has some fallback mechanism when FlurlHttpException happens.

**Comment by tmenier** [comment]
I did sort of a minimal fix for this - allowing you to do this without any issues:

```
new FlurlHttpException(null);
```

A `FlurlHttpException` would never be created without a non-null `FlurlCall` within the library, so I didn't really want to "encourage" this by providing an empty constructor, but hopefully this is a good compromise.

**Comment by tmenier** [comment]
Fixed & released in 3.0.1

**Comment by phatt-millips** [comment]
I ran into this issue also. I was trying to do `new FlurlHttpException(new FlurlCall());` in a Moq setup and got the same exception
